### PR TITLE
Make aspect ratio handling explicit

### PR DESCRIPTION
**Pull request by dnfield**
## Description

Adds a parameter to make the aspect ratio handling of image decoding explicit.

Today, if you specify both target width and height, the aspect ratio is not respected. If, on the other hand, you specify only target width or target height, the aspect ratio is maintained.

This is confusing for developers because they may not know the aspect ratio of the image bytes when using `instantiateImageCodec`, and they may want to specify maximum or minimum dimensions along those axes - e.g., to not allow the image to exceed screen resolution or widget size.

For now, to maintain backwards compatibility, this defaults to existing behavior. Eventually, this should be set to a default of `maintainLargest` so that we can have a more consistent model of how resizing is handled.

## Related Issues

https://github.com/flutter/flutter/issues/59578

## Tests

I added the following tests:

C++ and dart tests exercising the new enum values.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: testing/dart/encoding_test.dart [hunk 1]
```diff
@@ -114,7 +114,7 @@ class GrayscaleImage {
   GrayscaleImage._();
 
   static Future<Image> load() async {
-    final Uint8List bytes = await readFile('4x4.png');
+    final Uint8List bytes = await readFile('2x2.png');
```
Comment (chinmaygarde) [inline testing/dart/encoding_test.dart:117]: The file below seems unchanged. Did you forget to stage it?

Comment (chinmaygarde) [reply testing/dart/encoding_test.dart:117]: Hmm, the existing assertions seem unchanged. Maybe the file was incorrectly named before?

File: lib/ui/painting/image_decoder_unittests.cc [hunk 2]
```diff
@@ -571,12 +571,77 @@ TEST(ImageDecoderTest, VerifySimpleDecodingWithUpscaling) {
...
 
   ASSERT_EQ(
-      ImageFromCompressedData(data, 900, 300, ImageUpscalingMode::kAllowed,
+      ImageFromCompressedData(data, 900, 300, ImageUpscalingMode::kAllowed, {},
                               fml::tracing::TraceFlow(""))
           ->dimensions(),
       SkISize::Make(900, 300));
 }
 
+TEST(ImageDecoderTest, VerifySimpleDecodingAspectRatioNone) {
```
Comment (chinmaygarde) [inline lib/ui/painting/image_decoder_unittests.cc:580]: I don't believe we have used value parameterized tests in the engine before (I just found that gtest could do this recently) but this seems like a good place to use them. This pattern is pretty readable too. So, don't a strong opinion either way.

File: lib/ui/painting.dart [hunk 3]
```diff
@@ -1723,6 +1754,9 @@ Future<Codec> instantiateImageCodec(
   int? targetWidth,
   int? targetHeight,
   bool allowUpscaling = true,
+  // TODO(dnfield): Default this to maintainLargest once upstream tests have
+  // been updated, https://github.com/flutter/flutter/issues/59578
+  AspectRatioConstraint? targetAspectRatioConstraint,
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting.dart:1759]: We could have done that if we didn't already sometimes scale and sometimes not. My goal is to eventually make this non-optional after updating the framework side tests to explicitly use this parameter.

I suppose we could still do that if we document that if you set one and only one dimension, we ignore the aspect ratio constraint. I'm not sure if that's more confusing or less - it should at least be an error to use `.none` with either side unset.

File: lib/ui/painting/image_decoder_unittests.cc [hunk 4]
```diff
@@ -533,7 +533,7 @@ TEST(ImageDecoderTest, VerifySimpleDecoding) {
   ASSERT_EQ(SkISize::Make(600, 200), image->dimensions());
 
   ASSERT_EQ(ImageFromCompressedData(data, 6, 2, ImageUpscalingMode::kNotAllowed,
-                                    fml::tracing::TraceFlow(""))
+                                    {}, fml::tracing::TraceFlow(""))
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/image_decoder_unittests.cc:536]: This is just temporary until we make the input parameter non-null. Once that's the case here, I'll use an explicit value. Can switch to nullopt for now.

File: testing/dart/encoding_test.dart (hunk 1 above)
Comment (dnfield) [reply testing/dart/encoding_test.dart:117]: The file is actually a 2x2 png, not 4x4. I renamed the file and updated the strings.

File: lib/ui/painting.dart [hunk 5]
```diff
@@ -1695,6 +1695,31 @@ class Codec extends NativeFieldWrapperClass2 {
...
+///
+/// This enum is used in conjunction with the `targetWidth` and `targetHeight`
+/// parameters to [instantiateImageCodec].
+// Keep in sync with AspectRatioConstraint in image_decoder.h
+enum AspectRatioConstraint {
+  /// Do not maintain aspect ratio. This will cause the image to skew or stretch
+  /// if the target dimensions to not match its intrinsic dimensions.
+  none,
+
+  /// Maintain aspect ratio by keeping the target width of the image in tact.
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting.dart:1708]: I think we should throw in that case actually - it seems like it's a mistake to say you want to preserve the width in your target dimension and then not specify a target width.

**Comment by dnfield** [comment]
I'm wondering if an alternative to this patch is to just provide a new interface on the engine that lets you get the image descriptor. We could then update the framework to use that API instead, and avoid continuing to complect this interface. I had been avoiding that but I'm no longer sure why I should avoid it - except for the fact that I have no idea how it would be implemented for web (/cc @yjbanov @ferhatb)

**Comment by chinmaygarde** [comment]
Yeah, the original API was simple and we tacked on all these things it wasn't meant to support from the start. How about given a a buffer, return the intrinsic dimensions, EXIF information and other details to framework which then tells the engine exactly what it wants to do. I would be nervous about this approach if any of this depends on information only the engine knows. But this is not the case and doing it purely in Dart code in the framework seems like the way to go. I'd prefer keeping the logic to read image descriptor and the actual image decompression in the engine however. We can delegate all that to Skia or other platform specific codecs.

**Comment by chinmaygarde** [comment]
> I'd prefer keeping the logic to read image descriptor and the actual image decompression in the engine however...

On potential issue with this approach is that accessing the large buffers in the Dart VMs may require copies. So the API would need to account for ways to eliminate these copies. Copying large buffers twice (once to read the descriptor and once again to decode) would be unfortunate.

**Comment by dnfield** [comment]
I'm going to look into rewriting some API here first. Will revive this if it's needed.

**Pull request closed without merging**